api/server.go

```diff
@@ -674,6 +676,376 @@ func (s *server) walletsFundSFHandler(jc jape.Context) {
+	if err != nil {
+		jc.Error(fmt.Errorf("failed to select siacoin elements: %w", err), http.StatusInternalServerError)
+		return
+	}
+	for _, sce := range sces {
+		locked = append(locked, types.Hash256(sce.ID))
+	}
+
+	if !siacoinChange.IsZero() {
+		if wcr.ChangeAddress == types.VoidAddress {
+			jc.Error(errors.New("change address must be specified"), http.StatusBadRequest)
+			return
+		}
+
+		wcr.Siacoins = append(wcr.Siacoins, types.SiacoinOutput{
+			Value:   siacoinChange,
+			Address: wcr.ChangeAddress,
+		})
+	}
+
+	sfes, sfBasis, siafundChange, err := s.wm.SelectSiafundElements(walletID, siafundInput)
+	if err != nil {
+		jc.Error(fmt.Errorf("failed to select siafund elements: %w", err), http.StatusInternalServerError)
+		return
+	}
+	for _, sfe := range sfes {
+		locked = append(locked, types.Hash256(sfe.ID))
+	}
+
+	if siafundChange > 0 {
+		if wcr.ChangeAddress == types.VoidAddress {
+			jc.Error(errors.New("change address must be specified"), http.StatusBadRequest)
+			return
+		}
+
+		wcr.Siafunds = append(wcr.Siafunds, types.SiafundOutput{
+			Value:   siafundChange,
+			Address: wcr.ChangeAddress,
+		})
+	}
+
+	knownAddresses := make(map[types.Address]wallet.Address)
+	getAddress := func(addr types.Address) (wallet.Address, error) {
+		if a, ok := knownAddresses[addr]; ok {
+			return a, nil
+		}
+		a, err := s.wm.WalletAddress(walletID, addr)
+
+		if err != nil {
+			return wallet.Address{}, err
+		}
+		knownAddresses[addr] = a
+		return a, nil
+	}
+
+	resp := WalletConstructV2Response{
+		Basis:        basis,
+		EstimatedFee: fee,
+	}
+
+	txn := types.V2Transaction{
+		MinerFee:       fee,
+		SiacoinInputs:  make([]types.V2SiacoinInput, 0, len(sces)),
+		SiacoinOutputs: wcr.Siacoins,
+		SiafundInputs:  make([]types.V2SiafundInput, 0, len(sfes)),
+		SiafundOutputs: wcr.Siafunds,
+	}
+
+	for _, sfe := range sfes {
+		addr, err := getAddress(sfe.SiafundOutput.Address)
+		if err != nil {
+			jc.Error(fmt.Errorf("failed to get address: %w", err), http.StatusInternalServerError)
+			return
+		}
+
+		sfi := types.V2SiafundInput{
+			Parent:       sfe,
+			ClaimAddress: wcr.ChangeAddress,
+		}
+
+		if addr.SpendPolicy != nil {
+			// best effort to fill spend policy
+			sfi.SatisfiedPolicy = types.SatisfiedPolicy{
+				Policy: *addr.SpendPolicy,
+			}
+		}
+		txn.SiafundInputs = append(txn.SiafundInputs, sfi)
+	}
+
+	if len(sfes) > 0 && basis != sfBasis {
+		txnset, err := s.cm.UpdateV2TransactionSet([]types.V2Transaction{txn}, sfBasis, basis)
+		if err != nil {
+			jc.Error(fmt.Errorf("failed to update transaction set: %w", err), http.StatusInternalServerError)
+			return
+		}
+		txn = txnset[0]
+	}
+
+	for _, sce := range sces {
+		addr, err := getAddress(sce.SiacoinOutput.Address)
+		if err != nil {
+			jc.Error(fmt.Errorf("failed to get address: %w", err), http.StatusInternalServerError)
+			return
+		}
+
+		sci := types.V2SiacoinInput{
+			Parent: sce,
+			SatisfiedPolicy: types.SatisfiedPolicy{
+				Policy: *addr.SpendPolicy,
```

Comment:
You are setting the policy here but then check if it is not nil and set it again